src/psum.c

```diff
@@ -0,0 +1,1300 @@
+          outp[i].i = xjp[i & mask] == NA_INTEGER ? NA_REAL : 0;
+        }
+      } break;
+      case REALSXP: {
+        double *xjp = REAL(xj);
+        for (int i=0; i<n; i++) {
+          outp[i].r = xjp[i & mask];
+          outp[i].i = ISNAN(xjp[i & mask]) ? NA_REAL : 0;
+        }
+      } break;
+      case CPLXSXP: {
+        xjp = COMPLEX(xj);
+        for (int i=0; i<n; i++) {
+          outp[i].r = xjp[i & mask].r;
+          outp[i].i = xjp[i & mask].i;
+        }
+      } break;
+      default:
+        error(_("Internal error: should have caught non-INTSXP/REALSXP input by now")); // # nocov
+      }
+      for (int j=1; j<J; j++) {
+        xj = VECTOR_ELT(x, j);
+        nj = LENGTH(xj);
+        const int mask = nj == 1 ? 0 : INT_MAX;
+        switch (TYPEOF(xj)) {
+        case LGLSXP: case INTSXP: {
+          int *xjp = INTEGER(xj);
+          for (int i=0; i<n; i++) {
+            if (!ISNAN_COMPLEX(outp[i])) {
+              if (xjp[i & mask] == NA_INTEGER) {
+                outp[i].r = NA_REAL;
+                outp[i].i = NA_REAL;
+              } else {
+                outp[i].r += xjp[i & mask];
+              }
+            }
+          }
+        } break;
+        case REALSXP: {
+          double *xjp = REAL(xj);
+          for (int i=0; i<n; i++) {
+            if (!ISNAN_COMPLEX(outp[i])) {
+              if (ISNAN(xjp[i & mask])) {
+                outp[i].r = NA_REAL;
+                outp[i].i = NA_REAL;
+              } else {
+                outp[i].r += xjp[i & mask];
+              }
+            }
+          }
+        } break;
+        case CPLXSXP: {
+          xjp = COMPLEX(xj);
+          for (int i=0; i<n; i++) {
+            if (!ISNAN_COMPLEX(outp[i])) {
+              if (ISNAN_COMPLEX(xjp[i & mask])) {
+                outp[i].r = NA_REAL;
+                outp[i].i = NA_REAL;
+              } else {
+                outp[i].r += xjp[i & mask].r;
+                outp[i].i += xjp[i & mask].i;
+              }
+            }
+          }
+        } break;
+        default:
+          error(_("Internal error: should have caught non-INTSXP/REALSXP input by now")); // # nocov
+        }
+      }
+    } break;
+    default:
+      error(_("Internal error: should have caught non-INTSXP/REALSXP input by now")); // # nocov
+    }
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+SEXP pprod(SEXP x, SEXP narmArg) {
+  if (!isNewList(x))
+    error(_("Internal error: x must be a list")); // # nocov
+
+  SEXP xj;
+  int J=LENGTH(x);
+  if (J == 0) {
+    error(_("Empty input"));
+  } else if (J == 1) {
+    SEXP xj = VECTOR_ELT(x, 0);
+    if (TYPEOF(xj) == VECSXP) {
+      return pprod(xj, narmArg);
+    } else {
+      // na.rm doesn't matter -- input --> output
+      return xj;
+    }
+  }
+
+  if (!isLogical(narmArg) || LENGTH(narmArg)!=1 || LOGICAL(narmArg)[0]==NA_LOGICAL)
+    error(_("na.rm must be TRUE or FALSE"));
+
+  SEXPTYPE outtype = INTSXP;
```

Comment:
For the record I compared a version where all-`LGLSXP` inputs to `pprod` are re-routed to `pall`.

It's about 30% faster, but this difference is wiped when coercing to integer (a bit strange though, since I would think this coercion would be almost instantaneous for `LGLSXP->INTSXP`).

I guess `coerceVector` is `duplicate()`ing the input, but we don't need that -- just to change the `TYPEOF` (the storage is the same). Do we have any tool for that? @jangorecki

Reply:
There is a copy even there, I traced that recently as well. We don't have yet tool for that.